### ISSUE TITLE
Inconsistent Type Inference for Default Parameters between Methods and Functions

### ISSUE THREAD

**Issue by BrutalSimplicity** [microsoft/pyright#3744]
**Describe the bug**
It seems that pyright has inconsistent behavior when inferring the types from default parameters applied to a function and method.

**To Reproduce**

Take the following sample:

```python
from subprocess import run

def FunctionRunner(run=run):
    def handler(cmd: str):
        run(cmd)
    return handler

class ClassRunner:
    def __init__(self, run=run):
        self._run = run

    def __call__(self, cmd: str):
        self._run(cmd)
```

- The inferred signature of the `run` parameter for `FunctionRunner` is unknown

<img width="311" alt="image" src="https://user-images.githubusercontent.com/5298126/180775883-54fbfc39-b030-44ef-97e0-2296e78e4e70.png">

- Yet, when mimicking the same behavior using a class it binds the type definition to the parameter. And I am able to use the reference with proper type information.

<img width="909" alt="image" src="https://user-images.githubusercontent.com/5298126/180776005-04f90e04-8b87-4fd9-9693-0556467f5183.png">

**Expected behavior**

I'm not sure what the *expected* behavior should be, but **I would ❤️ if pyright would infer the function binding the same as the method.**

I can see the argument against this as the type parameter is not explicitly defined. But, it would be really nice if it inferred that from the default parameter as that is the only type information known about the parameter. This seems like the desired behavior in most cases 😅 .  I also think this behavior lines up well with this desired feature request #3684

**Comment by BrutalSimplicity** [comment]
Hi @erictraut, thanks for the quick reply!

Initially reinstalling the vscode python extension and install pyright globally seemed to work and the parameters were displaying the type information inferred from the default assignment. **But, then** everything seemed to revert to the same behavior as before, and I'm not really sure what to make of it. 😩 

I also added an additional example where **both fail to bind type information** when they are assigned as default parameters with the first invocation. I imagine this might be expected behavior, but thought that it would be able to resolve to the return type of the first invocation since pyright already knows the return types of the chain.

```python
from subprocess import run
from typing import List


def FunctionRunner(run=run):
    def handler(cmd: str):
        run(cmd)

    return handler


class ClassRunner:
    def __init__(self, run=run):
        self._run = run

    def __call__(self, cmd: str):
        self._run(cmd)


def SomeHandler(f_run=FunctionRunner(), c_run=ClassRunner()):
    def handler(args: List[str]):
        f_run(" ".join(args))
        c_run(" ".join(args))

    return handler
```

![2022-07-25 19 59 29](https://user-images.githubusercontent.com/5298126/180900207-e6391331-911c-41d6-98fa-4f786cb7b1ca.gif)


I dumped as much relevant information I could think of, but am happy to try to provide additional data.

I've also pushed the sample repository [here](https://github.com/BrutalSimplicity/pyright_issue_3744).

```
Darwin 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000 arm64
===========vscode version===========
1.69.2
3b889b090b5ad5793f524b5d1d39fda662b96a2a
arm64
======================
Python 3.10.5
======================
ms-python.python@2022.10.1
ms-python.vscode-pylance@2022.7.40
======================
pyright 1.1.263
Loading configuration file at /Users/maestro/Development/sandbox/pyright-test/pyrightconfig.json
Assuming Python platform Darwin
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
venv .venv subdirectory not found in venv path /Users/maestro/Development/sandbox/pyright-test.
stubPath /Users/maestro/Development/sandbox/pyright-test/typings is not a valid directory.
Search paths for /Users/maestro/Development/sandbox/pyright-test
  /Users/maestro/.asdf/installs/nodejs/18.2.0/.npm/lib/node_modules/pyright/dist/typeshed-fallback/stdlib
  /Users/maestro/Development/sandbox/pyright-test
  /Users/maestro/Development/sandbox/pyright-test/typings
  /Users/maestro/.asdf/installs/nodejs/18.2.0/.npm/lib/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /Users/maestro/.asdf/installs/python/3.10.5/lib/python3.10
  /Users/maestro/.asdf/installs/python/3.10.5/lib/python3.10/lib-dynload
  /Users/maestro/.asdf/installs/python/3.10.5/lib/python3.10/site-packages
Searching for source files
Found 1 source file
pyright 1.1.263
0 errors, 0 warnings, 0 informations 
Completed in 0.454sec
```

**Comment by erictraut** [comment]
Thanks for the additional details. I was able to repro the problem. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.264, which I just published. It will also be included in a future release of pylance.
